src/rules/base-rule-builder.ts

```diff
@@ -164,6 +166,16 @@ export abstract class BaseRuleBuilder {
         return this;
     }
 
+    /**
+     * Match only requests when the callback returns true 
+     */
+    withCallback(
```

Comment:
I like everything here I think, except the name. It doesn't sound like it's matching something - `withCallback` sounds like it's attaching a callback that's going to do something.

I've previously discussed this same change (in #19) as `.matching`, like:

```javascript
server.get('/').matching((request) => request.headers['content-type'] != 'text/plain').thenReply(200);
```

How does that sound to you? It's slightly inconsistent with the other `withX` methods, but there's inconsistent names for matchers already (`forHost`) and I'm not sure trying to force everything to fit with a 'with' prefix as we keep adding new matchers is worthwhile. What do you think?